File: 2021_04_18_make_a_lisp_in_csharp/Mal/Mal.Tests/EnvTests.cs

using Microsoft.VisualStudio.TestTools.UnitTesting;
using static Mal.EnvM;
using static Mal.Types;
using PowerFP;
using System;
using System.Linq;

namespace Mal.Tests
{
    [TestClass]
    public class EnvTests
    {
        [TestMethod]
        public void EnvTest()
        {
            var a = new Symbol("a", NilV);
            var b = new Symbol("b", NilV);
            var c = new Symbol("c", NilV);
            var d = new Symbol("d", NilV);

            var env1 = new Env(
                MapM.Empty<Symbol, MalType>().Add(a, new Str("a")).Add(b, new Str("b")),
                null);

            var env2 = new Env(
                MapM.Empty<Symbol, MalType>().Add(c, new Str("c")),
                env1);


            env2.Set(d, new Str("d"));

            Assert.AreSame(env1, env2.Find(a));
            Assert.AreSame(env1, env2.Find(b));
            Assert.AreSame(env2, env2.Find(c));
            Assert.AreSame(env2, env2.Find(d));
            Assert.IsNull(env2.Find(new Symbol("e", NilV)));

            Assert.AreEqual(new Str("a"), env2.Get(a));
            Assert.AreEqual(new Str("b"), env2.Get(b));
            Assert.AreEqual(new Str("c"), env2.Get(c));
            Assert.AreEqual(new Str("d"), env2.Get(d));
            Assert.ThrowsException<Exception>(() => env2.Get(new Symbol("e", NilV)));
        }
    }
}

File: 2021_10_21_make_a_lisp_in_fsharp/MiniMal/MiniMal.Tests/EnvTests.fs
namespace MiniMal.Tests

open Microsoft.VisualStudio.TestTools.UnitTesting
open Types
open Env
open TestUtils

[<TestClass>]
type EnvTests() =

    [<TestMethod>]
    member this.EnvTest() =
        let env1 =
            Env(Map([ ("a", Str("a")); ("b", Str("b")) ]), None)
        let env2 = Env(Map([ ("c", Str("c")) ]), Some env1)
        env2.Set "d" (Str("d")) |> ignore

        Assert.AreSame(env1, (env2.Find "a").Value)
        Assert.AreSame(env1, (env2.Find "b").Value)
        Assert.AreSame(env2, (env2.Find "c").Value)
        Assert.AreSame(env2, (env2.Find "d").Value)
        Assert.IsTrue((env2.Find "e") = None)

        assertMalEquals (Str("a")) (env2.Get "a")
        assertMalEquals (Str("b")) (env2.Get "b")
        assertMalEquals (Str("c")) (env2.Get "c")
        assertMalEquals (Str("d")) (env2.Get "d")
        assertThrowsException (fun () -> env2.Get "e")
